Fix C_ARG_REGS/C_ARG_OPNDS confusion; port the defined YARV opcode, which requires them

noahgibbs opened the pull request.

noahgibbs committed:
Port the YJIT defined opcode; fix C_ARG_REGS

noahgibbs commented:
Huh. This is failing in an odd way on x86. I've marked it DRAFT - please don't merge yet.

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -2156,26 +2157,19 @@ fn gen_defined(
-    mov(cb, RAX, uimm_opnd(Qnil.into()));
-    cmovnz(cb, RAX, REG1);
+    asm.cmp(def_result, Opnd::Imm(0));
+    let out_value = asm.csel_nz(pushval.into(), Qnil.into());
```
I think this could be the cause of the failure.

The previous code was comparing AL (lowest 8-bit) against 0. The new comparison is comparing the whole 64-bit. The problem is that C booleans only use the lowest 8 bits, so we can't be sure what the higher bits are. Possibly we'd want to implement an `Opnd::sub_reg(num_bits)` method for the `Opnd` struct (this method already exists for x86 and arm).

Before proceeding with this, I'd print out the values of `def_result`, check if the higher bits are ever nonzero. There's a `print_int(asm, opnd)` function that you can use to print out `def_result`.

maximecb commented:
I just rebased `yjit_backend_ir` on upstream_ruby/master.

Apologies for the unintentional merge conflicts. You may need to rebase your branch on `yjit_backend_ir`.

noahgibbs commented:
Thanks! I'll try those things.

noahgibbs commented:
Okay, rebased onto yjit_backend_ir. No trouble there.

Added the print for the result of rb_vm_defined(). Yup, the results definitely have some higher bits set. So yeah, I'll need to just compare the lowest 8 bits. I'll see if exposing that is easy for this PR.

noahgibbs commented:
Ah, but that's not the whole problem. If I AND the result with 256, I still get different results on x86 and ARM. So something else is wrong. I'll keep looking.

maximecb commented:
> Ah, but that's not the whole problem. If I AND the result with 256, I still get different results on x86 and ARM. So something else is wrong. I'll keep looking.

This might be something where there's an ABI difference between how booleans are passed along on x86-64 and ARM64.

256 is the 9th bit though.

If we bitwise AND the result with 255, we should have that the value is nonzero for boolean true on both x86 and arm?

noahgibbs commented:
Oh hey, you're totally right. It's caffeine tolerance-reset week, so I'll claim that as an excuse.

When that's fixed... we're getting the same result from rb_vm_defined on both now, though I still have some kind of bug because I'm getting a different final result. But that should make it easier to find.

noahgibbs commented:
Nope, never mind, that was it. I had a problem left from debug code. With that fixed, this works. I'll get this PR finalised later today or tomorrow.

noahgibbs commented:
Okay, working for me now. Working version pushed to GitHub with the "AND" to do it at runtime. It's not as clear to me how to fix it in the backend by requesting a specific number of bits, though I found several ways that look like they won't work :-)

maximecb commented on yjit/src/backend/x86_64/mod.rs:
```diff
@@ -26,6 +26,15 @@ pub const _C_ARG_OPNDS: [Opnd; 6] = [
+    X86Opnd::Reg(RCX_REG),
+    X86Opnd::Reg(R8_REG),
+    X86Opnd::Reg(R9_REG)
+];
```
Maybe we could avoid defining these twice, use C_ARG_REGS[idx].unwrap_reg() when the reg is needed.

noahgibbs replied:
Given that we only seem to use them *outside* this file in gen_entry_prologue (and only two of them), it's possible we don't need them X86Opnd::Reg form of them at all.

For now I'll do the unwrap/rewrap, but the longer-term answer may be "export less" and then we can just keep it in this file's preferred representation.

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -2156,26 +2157,20 @@ fn gen_defined(
-    mov(cb, RAX, uimm_opnd(Qnil.into()));
-    cmovnz(cb, RAX, REG1);
+    let bool_result = asm.and(def_result, Opnd::UImm(255));
+    asm.cmp(bool_result, Opnd::Imm(0));
```
Here this can be slightly more efficient by using the `asm.test()` instruction. It basically does exactly the combination of `and` and `cmp` together, it's meant to test flag bits.

noahgibbs replied:
Ah, good point.

noahgibbs commented:
Okay. I think everything should be updated according to the comments here.

maximecb commented:
Well done Noah! :)
